internal/handlers/common/aggregations/group.go

```diff
@@ -0,0 +1,346 @@
+var accumulators = map[string]newAccumulatorFunc{
+	// sorted alphabetically
+	"$count": newCountAccumulator,
+}
+
+// groupStage represents $group stage.
+//
+//	{ $group: {
+//		_id: <groupExpression>,
+//		<groupBy[0].outputField>: {<groupBy[0].accumulator>: <groupBy[0].expression>},
+//		...
+//		<groupBy[N].outputField>: {<groupBy[N].accumulator>: <groupBy[N].expression>},
+//	}}
+type groupStage struct {
+	groupExpression any
+	groupBy         []groupBy
+}
+
+// groupBy represents accumulation to apply on the group.
+type groupBy struct {
+	expression  any
+	accumulate  func(ctx context.Context, in []*types.Document) (any, error)
+	outputField string
+}
+
+// newGroup creates a new $group stage.
+func newGroup(stage *types.Document) (Stage, error) {
+	fields, err := common.GetRequiredParam[*types.Document](stage, "$group")
+	if err != nil {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrStageGroupInvalidFields,
+			"a group's fields must be specified in an object",
+			"$group",
+		)
+	}
+
+	if fields.Len() == 0 {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrStageGroupMissingID,
+			"a group specification must include an _id",
+			"$group",
+		)
+	}
+
+	var groupKey any
+	var groups []groupBy
+
+	iter := fields.Iterator()
+
+	defer iter.Close()
+
+	for {
+		field, v, err := iter.Next()
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if field == "_id" {
+			groupKey = v
+
+			idAccumulator := idAccumulator{
+				expression: v,
+			}
+			groups = append(groups, groupBy{
+				outputField: field,
+				accumulate:  idAccumulator.Accumulate,
+				expression:  v,
+			})
+
+			continue
+		}
+
+		accumulation, ok := v.(*types.Document)
+		if !ok || accumulation.Len() == 0 {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrStageGroupInvalidField,
+				fmt.Sprintf("The field '%s' must be an accumulator object", field),
+				"$group",
+			)
+		}
+
+		// document contains only one.
+		if accumulation.Len() > 1 {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrStageGroupOneAccumulator,
+				fmt.Sprintf("The field '%s' must specify one accumulator", field),
+				"$group",
+			)
+		}
+
+		operator, expression, err := accumulation.Iterator().Next()
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		newAccumulator, ok := accumulators[operator]
+		if !ok {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrNotImplemented,
+				"unimplemented",
+				"$group",
+			)
+		}
+
+		accumulator, err := newAccumulator(accumulation)
+		if err != nil {
+			return nil, err
+		}
+
+		groups = append(groups, groupBy{
+			outputField: field,
+			accumulate:  accumulator.Accumulate,
+			expression:  expression,
+		})
+	}
+
+	if groupKey == nil {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
```

Comment:
I see codecov complains about this part, is there a reason why we don't have it tested?

Reply:
Thanks, there was another check a bit earlier in the code, I removed that so we don't have duplicated check!